### PR TITLE
Slightly improve diagnostics and remove assertions

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Unfortunately there are too much cases when these assertions can fail in expected way, for example: https://pastila.nl/?018012bd/646734aaab6fa02444b562cef4b07abe. Related to https://github.com/ClickHouse/ClickHouse/issues/41013

### CODE REVIEW

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
We should also set `temporary_directories_lifetime` to 1 second in Integration Tests and decrease default value of this setting to 6 or even 3 hours.

File: src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp [hunk 1]
```diff
@@ -56,7 +56,7 @@ void throwIfError(const Aws::Utils::Outcome<Result, Error> & response)
     if (!response.IsSuccess())
     {
         const auto & err = response.GetError();
-        throw Exception(ErrorCodes::S3_ERROR, "{} (Code: {})", err.GetMessage(), static_cast<size_t>(err.GetErrorType()));
+        throw S3Exception(fmt::format("{} (Code: {})", err.GetMessage(), static_cast<size_t>(err.GetErrorType())), err.GetErrorType());
```
Comment (tavplubix) [inline src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp:59]: Consider adding variadic ctor that calls `fmt::format` internally like in `DB::Exception`

**Comment by tavplubix** [comment]
Integration tests (release) [2/2] - `test_merge_tree_s3/test.py::test_s3_disk_reads_on_unstable_connection` - looks suspicious and probably related to the change in `ReadBufferFromS3`

**Review by tavplubix — APPROVED** [review]

**Comment by alesapin** [comment]
I really want to merge it.

**Pull request merged**